DeepFried2/layers/BackwardsConvolutionCUDNN.py

```diff
@@ -3,50 +3,61 @@
 from theano.sandbox.cuda.basic_ops import gpu_contiguous, gpu_alloc_empty
 from theano.sandbox.cuda import dnn
 
+import operator
+import functools
+
 
 class BackwardsConvolutionCUDNN(df.Module):
-    def __init__(self, n_input_plane, n_output_plane, k_w, k_h, d_w=1, d_h=1, pad_w=0, pad_h=0, mode='cross', with_bias=True, initW=df.init.xavier(), initB=df.init.const(0), border=None):
+    def __init__(self, n_input_plane, n_output_plane, filter_size, stride=(1,1), padding=(0,0), mode='cross', with_bias=True, initW=df.init.xavier(), initB=df.init.const(0), border=None):
         # mode='cross' is the default in Lasagne[1], Torch[2], matConvNet[3], Caffee[4].
         #
         # 1: https://github.com/Lasagne/Lasagne/blob/63d44a0d/lasagne/layers/dnn.py#L299
         # 2: https://github.com/soumith/cudnn.torch/blob/840f0228/SpatialConvolution.lua#L83
         # 3: https://github.com/vlfeat/matconvnet/blob/b7dd9c96/matlab/src/bits/impl/nnconv_cudnn.cu#L133
         # 4: https://github.com/BVLC/caffe/blob/50ab52cb/include/caffe/util/cudnn.hpp#L104
-        #
-        # `border` is an alternative way to specify `pad_w` and `pad_h` so that Theano strings can be used. Better documentation to follow soon.
+
         df.Module.__init__(self)
         self.n_input_plane = n_input_plane
         self.n_output_plane = n_output_plane
-        self.k_w = k_w
-        self.k_h = k_h
-        self.d_w = d_w
-        self.d_h = d_h
+        self.filter_size = filter_size
+        self.stride = stride
         self.mode = mode
         self.with_bias = with_bias
 
+        assert len(self.stride) == len(self.filter_size), 'The dimensionality of the stride and the filter size should match.'
         # 'same' is a (common) shortcut for "zero-padding so that outshape == inshape".
-        self.border = border or (pad_h, pad_w)
+        self.border = border or padding
         if self.border == 'same':
-            assert self.k_w % 2 == 1 and self.k_h % 2 == 1, "'same' convolution only supports odd filter sizes."
-            self.border = ((self.k_h-1)//2, (self.k_w-1)//2)
+            assert [k % 2 == 1 for k in self.filter_size] == [True] * len(self.filter_size), "'same' convolution only supports odd filter sizes."
+
+            self.border = tuple( (k - 1)//2 for k in self.filter_size )
 
-        w_shape = (n_input_plane, n_output_plane, k_h, k_w)
-        w_fan = (n_output_plane*k_w*k_h, n_input_plane*k_w*k_h)
+        assert len(self.border) == len(self.stride), 'The dimensionality of the stride and the padding should match.'
 
-        self.weight, self.grad_weight = create_param_and_grad(w_shape, initW, fan=w_fan, name='Wconv_{},{}@{}x{}'.format(n_input_plane, n_output_plane, k_w, k_h))
+        w_shape = (n_input_plane, n_output_plane) + self.filter_size
+        w_fan = (functools.reduce(operator.mul, self.filter_size, n_output_plane),
+                 functools.reduce(operator.mul, self.filter_size, n_input_plane))
```

Comment:
It's a little more straightforward to use `np.prod(self.filter_size)*n_output_plane`